src/game/client/draw/w3dtankdraw.cpp

```diff
@@ -0,0 +1,379 @@
+/**
+ * @file
+ *
+ * @author Jonathan Wilson
+ *
+ * @brief W3D Tank Draw Module
+ *
+ * @copyright Thyme is free software: you can redistribute it and/or
+ *            modify it under the terms of the GNU General Public License
+ *            as published by the Free Software Foundation, either version
+ *            2 of the License, or (at your option) any later version.
+ *            A full copy of the GNU General Public License can be found in
+ *            LICENSE
+ */
+#include "w3dtankdraw.h"
+#include "aiupdate.h"
+#include "drawable.h"
+#include "mapper.h"
+#include "matinfo.h"
+#include "object.h"
+#include "particlesys.h"
+#include "particlesysmanager.h"
+#include "physicsupdate.h"
+#include "scriptengine.h"
+#include "vertmaterial.h"
+#include "view.h"
+#include <cstring>
+
+using std::strchr;
+
+W3DTankDrawModuleData::W3DTankDrawModuleData() :
+    m_treadDebrisNameLeft("TrackDebrisDirtLeft"),
+    m_treadDebrisNameRight("TrackDebrisDirtRight"),
+    m_treadAnimationRate(0.0f),
+    m_treadPivotSpeedFraction(0.6f),
+    m_treadDriveSpeedFraction(0.3f)
+{
+}
+
+void W3DTankDrawModuleData::Build_Field_Parse(MultiIniFieldParse &p)
+{
+    // clang-format off
+    static FieldParse dataFieldParse[] = {
+        { "TreadDebrisLeft", &INI::Parse_AsciiString, nullptr, offsetof(W3DTankDrawModuleData, m_treadDebrisNameLeft) },
+        { "TreadDebrisRight", &INI::Parse_AsciiString, nullptr, offsetof(W3DTankDrawModuleData, m_treadDebrisNameRight) },
+        { "TreadAnimationRate", &INI::Parse_Velocity_Real, nullptr, offsetof(W3DTankDrawModuleData, m_treadAnimationRate) },
+        { "TreadPivotSpeedFraction", &INI::Parse_Real, nullptr, offsetof(W3DTankDrawModuleData, m_treadPivotSpeedFraction) },
+        { "TreadDriveSpeedFraction", &INI::Parse_Real, nullptr, offsetof(W3DTankDrawModuleData, m_treadDriveSpeedFraction) },
+        { nullptr, nullptr, nullptr, 0 },
+    };
+    // clang-format on
+
+    W3DModelDrawModuleData::Build_Field_Parse(p);
+    p.Add(dataFieldParse, 0);
+}
+
+W3DTankDraw::W3DTankDraw(Thing *thing, ModuleData const *module_data) :
+    W3DModelDraw(thing, module_data),
+    m_treadDebrisLeft(nullptr),
+    m_treadDebrisRight(nullptr),
+    m_prevRenderObj(nullptr),
+    m_treadCount(0)
+{
+    for (int i = 0; i < MAX_TREADS_PER_TANK; i++) {
+        m_treads[i].m_robj = nullptr;
+    }
+
+    m_treadDirection.x = 1.0f;
+    m_treadDirection.y = 0.0f;
+    m_treadDirection.z = 0.0f;
+    Create_Emitters();
+}
+
+W3DTankDraw::~W3DTankDraw()
+{
+    for (int i = 0; i < MAX_TREADS_PER_TANK; i++) {
+        Ref_Ptr_Release(m_treads[i].m_robj);
+    }

```

Comment:
Looks like the emitters are leaked here. Should this call `Toss_Emitters` ?

Reply:
I can't see anything else that would free these emitters and W3DTruckDraw's destructor calls W3DTruckDraw::Toss_Emitters so it seems logical to add a call here. fixed.